_i18n/en/_docs/stalker.md

```diff
@@ -0,0 +1,1859 @@
+
+Finally, if a block of code ends with a deterministic branch (e.g. the destination is fixed and the branch is not conditional) then rather than replacing that last branch with a call back to Stalker to instrument the next block, we can instrument the next block ahead of time and direct control flow there without having to re-enter the stalker engine. This process is called backpatching. In actual fact, we can deal with conditional branches too, if we instrument both blocks of code (the one if the branch is taken and the one if it isn't) then we can replace the original conditional branch with one conditional branch which directs control flow to instrumented version of the block encountered when the branch was taken, followed by a unconditional branch to the other instrumented block. We can also deal partially with branches where the target is not static. Say our branch is something like:
+
+```
+BR x0
+```
+
+This sort of instruction is common when calling a function pointer, or class method. Whilst the value of x0 can change, quite often it will actually always be the same. In this case, we can replace the final branch instruction with code which compares the value of x0 against our known function, and if it matches branches to the address of the instrumented copy of the code. This can then be followed by an unconditional branch back to the stalker engine if it doesn't match. So if the value of the function pointer say is changed, then the code will still work and we will re-enter stalker and instrument wherever we end up, however, if as we expect it remains unchanged then we can bypass the stalker engine altogether and go straight to the instrumented function.
+
+
+## Options
+Now lets look at the options when we follow a thread with stalker. Stalker generates events when a followed thread is being executed, these are placed onto a queue and flushed either periodically or when the queue is full. The size and time period can be configured by the options. Events can be generated for on a per-instruction basis either for calls, returns or all instructions. Or they can be generated on a block basis, either when a block is executed, or when it is instrumented by the stalker engine. The difference here is that blocks may be instrumented before they are executed as an optimization by Stalker. Recall that when a call (`BL`, `BLX` etc) instruction is encountered, the callee and the block of the caller following the instruction is instrumented as we can be pretty sure it will run. This may be useful for code covereage.
```

Comment:
How does a compile event differ from the event and block events? Presumably, it is when the block is instrumented, rather then when it is run? What circumstances are there where you may instrument a block, but then not execute it (so that one event doesn't occur immediately after the other)?

Cheers.

Reply:
> How does a compile event differ from the event and block events?

The `compile` event is emitted right after a basic block has been compiled, i.e. just before it's executed for the first time. So it's only done once (per thread). This makes it very useful for measuring coverage. (However, if the code mutates it will happen again for the same basic block.)

The `block` event is emitted each time a given basic block is executed, *before* any of its instructions have been executed.

> Presumably, it is when the block is instrumented, rather then when it is run?

Correct.

> What circumstances are there where you may instrument a block, but then not execute it (so that one event doesn't occur immediately after the other)

None that I can think of.

Cheers!